src/miner.erl

```diff
@@ -750,75 +801,166 @@ common_enough_or_default(Threshold, Xs, Default) ->
         [{_, _}|_]                     -> Default % Not common-enough.
     end.
 
-set_next_block_timer(State=#state{blockchain=Chain}) ->
+-spec calculate_next_block_time(blockchain:blockchain()) -> integer().
+calculate_next_block_time(Chain) ->
+    calculate_next_block_time(Chain, false).
+
+-spec calculate_next_block_time(blockchain:blockchain(), boolean()) -> integer().
+calculate_next_block_time(Chain, Subsequent) ->
     Now = erlang:system_time(seconds),
-    {ok, BlockTime0} = blockchain:config(?block_time, blockchain:ledger(Chain)),
-    {ok, #block_info_v2{time=LastBlockTime, height=Height}} = blockchain:head_block_info(Chain),
+    Ledger = blockchain:ledger(Chain),
+    {ok, BlockTime0} = blockchain:config(?block_time, Ledger),
     BlockTime = BlockTime0 div 1000,
+    {ok, #block_info_v2{time=LastBlockTime, height=Height}} = blockchain:head_block_info(Chain),
     LastBlockTimestamp = case Height of
                              1 ->
                                  %% make up a plausible time for the genesis block
                                  Now;
                              _ ->
                                  LastBlockTime
                          end,
-
-    StartHeight0 = application:get_env(miner, stabilization_period, 0),
-    StartHeight = max(1, Height - StartHeight0),
-    {ActualStartHeight, StartBlockTime} =
-        case Height > StartHeight of
-            true ->
-                case blockchain:find_first_block_after(StartHeight, Chain) of
-                    {ok, Actual, StartBlock} ->
-                        {Actual, blockchain_block:time(StartBlock)};
-                    _ ->
-                        {0, undefined}
-                end;
-            false ->
-                {0, undefined}
+    %% mimic snapshot_take functionality for block range window
```

Comment:
What's the benefit of looking at the very short ("snapshot_take") block time average here? I fear this is going to add a lot of volatility to block times (in addition to adding complexity to this code). I believe you are trying to catch up quickly. However, to what benefit? I think it would be preferable to catch up over a longer period of time to have less volatility in times from block to block. There is a high chance that this period includes multiple rewards blocks which the algorithm will try to quick compensate for which exceptionally quick in my opinion is unnecessary.